packages/@aws-cdk/core/lib/secret-value.ts

```diff
@@ -52,6 +59,11 @@ export class SecretValue extends Intrinsic {
     ];
 
     const dyref = new CfnDynamicReference(CfnDynamicReferenceService.SECRETS_MANAGER, parts.join(':'));
+
+    if (secretId.startsWith('arn:')) {
+      const secretArn = options.jsonField ? `${secretId}:${options.jsonField}` : secretId;
+      return this.cfnDynamicReference(dyref, secretArn);
```

Comment:
I don't like this extra return... let's re-structure the code slightly:

```ts
    const dyref = new CfnDynamicReference(CfnDynamicReferenceService.SECRETS_MANAGER, parts.join(':'));
    const secretQualifier: SecretQualifer = {
      secretId,
      jsonField: options.jsonField,
      versionStage: options.versionStage,
      versionId: options.versionId,
    };
    return new SecretValue(dyref, {
      secretQualifier,
    });
```

(We'll handle the parsing of ARN in the CodeBuild Project, thus re-using that logic)